[Tools][WPE] run-mvt-tests should retry unexpected failures

clopez opened the pull request:
#### 3e12a025ba711af6d7027624fbcc7cc2154ba9f3
<pre>
[Tools][WPE] run-mvt-tests should retry unexpected failures
<a href="https://bugs.webkit.org/show_bug.cgi?id=289349">https://bugs.webkit.org/show_bug.cgi?id=289349</a>

Reviewed by Adrian Perez de Castro.

Some of the MVT tests are randomly flaky, and this is causing issues
on the bot runners due this random unexpected failures.

This patch modifies the runner so each unexpected failure is retried
up to N times (default 3) before marking it as such.

It also improves the conditions to check when the tests have started
or finished. Previously there was a bug detecting that all the tests
finished completely. It also improves the logging to summarize at the
end of each step the names of the tests that failed (expected or not).

* Tools/Scripts/run-mvt-tests:
(parse_args):

Canonical link: <a href="https://commits.webkit.org/292058@main">https://commits.webkit.org/292058@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8e9a55516ed7562c70d25a2fb30f6e4c7988fb1a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94343 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13930 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3703 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99355 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44871 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/96393 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14231 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22360 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72001 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29336 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97345 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10577 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85206 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52323 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10266 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2886 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44189 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80512 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2987 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101400 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21395 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15618 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81004 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21647 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81212 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80362 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24917 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2308 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14628 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15225 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21379 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21066 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24526 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22807 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

spenap commented on Tools/Scripts/run-mvt-tests:
```diff
@@ -145,8 +146,14 @@ class MVTWebDriverRunner():
-        return driver.execute_script('return !(typeof globalRunner === "undefined" || globalRunner === null);')
+        return driver.execute_script('return globalRunner.hasOwnProperty("testToRun") && globalRunner.hasOwnProperty("currentTestIdx") && globalRunner.hasOwnProperty("testList");')
+
+    def _did_all_tests_finished(self, driver):
```
```suggestion
    def _have_all_tests_finished(self, driver):
```

spenap commented on Tools/Scripts/run-mvt-tests:
```diff
@@ -159,7 +166,7 @@ class MVTWebDriverRunner():
                 running_test_id = self.driver.execute_script("return globalRunner.currentTestIdx;")
                 total_tests_to_run = self.driver.execute_script("return globalRunner.testList.length;")
-                if total_tests_to_run > 0 and running_test_id==total_tests_to_run:
+                if total_tests_to_run > 0 and running_test_id==total_tests_to_run and self._did_all_tests_finished(self.driver):
```
```suggestion
                if total_tests_to_run > 0 and running_test_id == total_tests_to_run and self._have_all_tests_finished(self.driver):
```

spenap commented on Tools/Scripts/run-mvt-tests:
```diff
@@ -177,10 +184,40 @@ class MVTWebDriverRunner():
+            self.driver.get(test_url)
+            WebDriverWait(self.driver, 20).until(self._suite_has_loaded)
+            WebDriverWait(self.driver, 20).until(self._suite_has_started)
+            WebDriverWait(self.driver, 1000).until(self._did_all_tests_finished)
```
```suggestion
            WebDriverWait(self.driver, 1000).until(self._have_all_tests_finished)
```

spenap commented on Tools/Scripts/run-mvt-tests:
```diff
@@ -289,18 +326,45 @@ def run_mvt_test_suite(args):
+        tests_run = len(tests_run_list)
+        # Maybe retry unexpected failures
+        if args.retry_unexpected_failures > 0 and unexpected_failures > 0:
+            unexpected_failures_str = "', '".join(unexpected_failures_list)
```
why "', '"? What is the output expected to look like?

clopez replied:
Example of output:

```
Retrying 3 unexpected failures: 'DASH_DYNAMIC Playback', 'DASH_DYNAMIC Pause', 'DASH_MULTIAUDIO AudioTracks':
[Repeat 1/3] Running test "DASH_DYNAMIC Playback" from suite "dash-dashjs-test" at https://mvt.rdkcentral.com/?test_type=dash-dashjs-test&tests=9&command=run
Test result: failed
[Repeat 2/3] Running test "DASH_DYNAMIC Playback" from suite "dash-dashjs-test" at https://mvt.rdkcentral.com/?test_type=dash-dashjs-test&tests=9&command=run
Test result: failed
[Repeat 3/3] Running test "DASH_DYNAMIC Playback" from suite "dash-dashjs-test" at https://mvt.rdkcentral.com/?test_type=dash-dashjs-test&tests=9&command=run
Test result: failed
[Repeat 1/3] Running test "DASH_DYNAMIC Pause" from suite "dash-dashjs-test" at https://mvt.rdkcentral.com/?test_type=dash-dashjs-test&tests=22&command=run
Test result: failed
[Repeat 2/3] Running test "DASH_DYNAMIC Pause" from suite "dash-dashjs-test" at https://mvt.rdkcentral.com/?test_type=dash-dashjs-test&tests=22&command=run
Test result: failed
[Repeat 3/3] Running test "DASH_DYNAMIC Pause" from suite "dash-dashjs-test" at https://mvt.rdkcentral.com/?test_type=dash-dashjs-test&tests=22&command=run
Test result: failed
[Repeat 1/3] Running test "DASH_MULTIAUDIO AudioTracks" from suite "dash-dashjs-test" at https://mvt.rdkcentral.com/?test_type=dash-dashjs-test&tests=40&command=run
Test result: passed
[Repeat 2/3] Running test "DASH_MULTIAUDIO AudioTracks" from suite "dash-dashjs-test" at https://mvt.rdkcentral.com/?test_type=dash-dashjs-test&tests=40&command=run
Test result: passed
[Repeat 3/3] Running test "DASH_MULTIAUDIO AudioTracks" from suite "dash-dashjs-test" at https://mvt.rdkcentral.com/?test_type=dash-dashjs-test&tests=40&command=run
Test result: passed
[FLAKY][PASS] DASH_MULTIAUDIO AudioTracks.
```

spenap replied:
Thanks! I understand now why you wanted to separately quote each specific failure :-)

spenap commented on Tools/Scripts/run-mvt-tests:
```diff
@@ -289,18 +326,45 @@ def run_mvt_test_suite(args):
         _log.info(f"    - {tests_pass} tests passed.")
         if expected_failures > 0:
-            _log.info(f"    - {expected_failures} tests were expected failures.")
+            _log.info(f"    - {expected_failures} tests were expected failures:")
```
I think you could use https://docs.python.org/3/library/string.html#formatspec to add the padding / width instead

spenap reviewed (commented):
LGTM (but I'm not a reviewer)

I'd suggest to use a Python formatter such as [Black](https://black.readthedocs.io/en/stable/index.html), if none is currently enforced, unless the WebKit community has explicitly chosen style freedom with Python. This would make some of the style choices more consistent (e.g. spaces around operators, that kind of thing)

clopez committed:
[Tools][WPE] run-mvt-tests should retry unexpected failures
https://bugs.webkit.org/show_bug.cgi?id=289349

Reviewed by Adrian Perez de Castro.

Some of the MVT tests are randomly flaky, and this is causing issues
on the bot runners due this random unexpected failures.

This patch modifies the runner so each unexpected failure is retried
up to N times (default 3) before marking it as such.

It also improves the conditions to check when the tests have started
or finished. Previously there was a bug detecting that all the tests
finished completely. It also improves the logging to summarize at the
end of each step the names of the tests that failed (expected or not).

* Tools/Scripts/run-mvt-tests:
(parse_args):

Canonical link: https://commits.webkit.org/292058@main
